hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,739 @@
+    nrf_delay_us(10); // need a bit delay to switch between normal flash and security register
+    // Send address bytes, do not finalize long frame transfer
+    ret = nrfx_qspi_lfm_xfer(addrBuf, nullptr, sizeof(addrBuf), false);
+    CHECK_NRF_RETURN(ret, nrf_system_error(ret));
+    // Send data to write, finalize long frame transfer
+    ret = nrfx_qspi_lfm_xfer((const void *)buf, nullptr, size, true);
+    CHECK_NRF_RETURN(ret, nrf_system_error(ret));
+
+    return 0;
+}
+
+int gd25_security_register_operation(hal_exflash_gd25_security_register_read_write operation, uintptr_t addr, uint8_t* data_buf, size_t data_size) {
+    // Determine which OTP Register to start operating from
+    size_t otp_page = addr / HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25;
+    size_t data_to_process = data_size;
+    size_t data_processed = 0;
+    size_t data_to_process_this_operation = 0;
+    int ret = 0;
+
+    while (data_to_process && !ret) {
+        // Determine the offset of the current OTP register to start reading from
+        size_t register_offset = (addr + data_processed) % HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25;
+        data_to_process_this_operation = data_to_process;
+
+        // If the data remaining to read extends into the next register, truncate the read amount to fit into this register
+        if (register_offset + data_to_process_this_operation > HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25){
+            data_to_process_this_operation = HAL_QSPI_FLASH_OTP_SECTOR_SIZE_GD25 - register_offset;
+        }
+
+        ret = operation(otp_page, register_offset, &data_buf[data_processed], data_to_process_this_operation);
+
+        // Increment/decrement counter variables, increment to the next OTP register
+        data_to_process -= data_to_process_this_operation;
+        data_processed += data_to_process_this_operation;
+        otp_page++;
+    }
+
+    return ret;
+}
+
+int erase_common(uintptr_t start_addr, size_t num_blocks, nrf_qspi_erase_len_t len) {
+    FlashLock lk;
+
+    int err_code = NRF_SUCCESS;
+
+    const size_t block_length = len == NRF_QSPI_ERASE_LEN_4KB ? 4096 : 64 * 1024;
+
+    /* Round down to the nearest block */
+    start_addr = ((start_addr / block_length) * block_length);
+
+    for (size_t i = 0; i < num_blocks; i++) {
+        err_code = nrfx_qspi_erase(len, start_addr);
+        CHECK_NRF_RETURN(err_code, nrf_system_error(err_code));
+
+        exflash_qspi_wait_completion();
+
+        start_addr += block_length;
+    }
+
+    // LOG_DEBUG(TRACE, "Erased %lu %lukB blocks starting from %" PRIxPTR,
+    //           num_blocks, block_length / 1024, start_addr);
+
+    return 0;
+}
+
+} // Anonymous
+
+int hal_exflash_init(void) {
+    // PATCH: Initialize CS pin, external memory discharge
+    nrf_gpio_cfg_output(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_pin_set(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_cfg_input(QSPI_FLASH_IO1_PIN, NRF_GPIO_PIN_PULLDOWN);
+
+    FlashLock lk;
+
+    hal_qspi_flash_type_t all_flashes[] = {
```

Comment:
Nit: This should probably be coming from exflash_hal_params or even perhaps the platform config file?